packages/lib/gpcircuits/src/multituple.ts

```diff
@@ -0,0 +1,232 @@
+
+  // The tuple indices will begin with the first `paramTupleArity` indices.
+  const tupleIndices = [indices.slice(0, paramTupleArity)];
+
+  // If we have more than that,
+  if (indices.length > paramTupleArity) {
+    // split the remainder into `paramTupleArity - 1` sized chunks.
+    const unlinkedChunks = toChunks(
+      indices.slice(paramTupleArity),
+      paramTupleArity - 1
+    );
+
+    // Prepend the 'index' of the preceding tuple
+    // to each of these chunks and push them onto `tupleIndices`.
+    unlinkedChunks.forEach((chunk, i) =>
+      tupleIndices.push([firstAvailableTupleIndex + i].concat(chunk))
+    );
+  }
+
+  // Take note of the last tuple for padding purposes.
+  const lastTupleIdx = tupleIndices.length - 1;
+  const lastTuple = tupleIndices[lastTupleIdx];
+  const lastTupleLength = lastTuple.length;
+
+  // Pad the last tuple of `tupleIndices` with `firstIndex` if it is not of length `paramTupleArity` and return.
+  for (let i = 0; i < paramTupleArity - lastTupleLength; i++) {
+    lastTuple.push(firstIndex);
+  }
+
+  return tupleIndices;
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'tuple hash',
+ * thus chaining together smaller tuples to form/represent a larger one.
+ * It accomplishes this by hashing the values, then reducing the resulting
+ * tuple down to one element, `paramTupleArity` elements at a time, via
+ * the Poseidon hash function, padding the tuple with the hash of `elements[0]`
+ * if necessary. The precise mechanism is described in `multiTupleHasher` below.
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param elements the values forming the input tuple
+ * @returns appropriately-formed hash of the input tuple
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function hashTuple(
+  paramTupleArity: number,
+  elements: PODValue[]
+): bigint {
+  // Call the multituple hasher.
+  const multihash = multiTupleHasher(paramTupleArity, elements);
+
+  // Return the last hash.
+  return multihash[multihash.length - 1];
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'multituple hash',
+ * which is the result of (Poseidon) hashing the first `paramTupleArity` elements
+ * and subsequently hashing every successive chunk of `paramTupleArity` - 1 elements,
+ * prepending the previously computed hash and padding the remainder of the list
+ * with the first element to form a list of exactly `paramTupleArity` elements
+ * to feed the `tupleHasher`. All of these hashes are returned. This process
+ * mimics the behaviour of the `MultiTupleModule`, which chains together
+ * smaller tuples to form/represent a larger one.
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param elements the values forming the input tuple
+ * @returns the list of multituple hashes
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function multiTupleHasher(
```

Comment:
The structure of this function is sensible, though do you anticipate it being used anywhere outside of hashTuple?  I could see it being helpful to be able to calculate the expected output values of all the tuples in a module, but right now the code doesn't make use of that, instead making separate calls to hashTuple where necessary.

Reply:
I suppose not. Its main purpose is to check the output of the multituple module, since the only use we currently have for chains of tuples is reducing a longer tuple down to a single hash for list membership checks, in which case `hashTuple` suffices as you noted.

Reply:
That also begs the question of whether there are additional test cases to write based on that. :)

Fine to leave it as-is if it's potentially useful in future, particularly since it could come up as you implement the GPC config layer.